cranelift/codegen/src/isa/aarch64/abi.rs

```diff
@@ -1469,12 +1484,49 @@ impl ABICall for AArch64ABICall {
         from_reg: Reg,
     ) {
         match &self.sig.args[idx] {
-            &ABIArg::Reg(reg, ty) => ctx.emit(Inst::gen_move(
-                Writable::from_reg(reg.to_reg()),
-                from_reg,
-                ty,
-            )),
-            &ABIArg::Stack(off, ty) => {
+            &ABIArg::Reg(reg, ty, ext)
+                if ext != ir::ArgumentExtension::None && ty_bits(ty) < 64 =>
+            {
+                assert_eq!(RegClass::I64, reg.get_class());
+                let signed = match ext {
+                    ir::ArgumentExtension::Uext => false,
+                    ir::ArgumentExtension::Sext => true,
+                    _ => unreachable!(),
+                };
+                ctx.emit(Inst::Extend {
+                    rd: Writable::from_reg(reg.to_reg()),
+                    rn: from_reg,
+                    signed,
+                    from_bits: ty_bits(ty) as u8,
+                    to_bits: 64,
+                });
+            }
+            &ABIArg::Reg(reg, ty, _) => {
+                ctx.emit(Inst::gen_move(
+                    Writable::from_reg(reg.to_reg()),
+                    from_reg,
+                    ty,
+                ));
+            }
+            &ABIArg::Stack(off, ty, ext) => {
+                if ext != ir::ArgumentExtension::None && ty_bits(ty) < 64 {
```

Comment:
Presumably (correct me if I'm wrong) this is correct but a bit suboptimal, in that we're extending a value that we read from the stack in a previous insn; and in the ideal case we could have just done a suitable widening load?  Is that a correct understanding?  I'm not saying this should be fixed, but just checking I have this right.

Reply:
And ditto observation for the x64 equivalent of this, AFAICS.

Reply:
Other way around -- we're storing a value in a register to the stack; sadly there's no store instruction (as far as I know) that can sign/zero-extend the bottom N bits of a GPR into 64 bits and then store that 64 bits, all in one go.